### PR TITLE
tests: more specific apparmor check

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Why is this needed?

Fixes: #3071

## Test Steps

I temporarily applied this diff https://pastebin.canonical.com/p/68W67tpDYj/
That diff _inconsistently_ causes xenial tests to cause apparmor denials.

Then I ran the following:
```
tox run -e behave -- --logging-level warning --no-junit --no-color features/_version.feature -D check_version=1:1+devel -D releases=xenial,jammy -D machine_types=lxd-container,lxd-vm
```
It might take a couple tries because it doesn't consistently trigger the apparmor denials, but eventually you can get a run where the xenial lxd-container test fails and displays apparmor DENIED logs from your host that are specific to that xenial test container. And you can get a xenial lxd-vm test fail with DENIED logs from the xenial vm. The jammy tests will always pass even if the xenial tests fail during the same run.

## Checklist

 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] Changes here need to be documented, and this was done in: 

## Does this PR require extra reviews?

 - [ ] Yes
 - [x] No

### CODE REVIEW

File: features/environment.py [hunk 1]
```diff
@@ -498,22 +499,47 @@ def after_scenario(context, scenario):
...
+            )
+            sut.instance.pull_file("/var/log/syslog", syslog_dest)
+            with open(syslog_dest, "r") as syslog_fd:
+                syslog_messages = syslog_fd.readlines()
+            apparmor_denied = [
+                msg.strip()
+                for msg in syslog_messages
+                if ("DENIED" in msg and "ubuntu_pro_" in msg)
+            ]
+            return apparmor_denied
```
Comment (panlinux) [inline features/environment.py:532]: This bit won't work, unfortunately. apparmor denials in the case of a lxd container will not show up in the container, but only in the host.

Comment (panlinux) [reply features/environment.py:532]: In other words, the apparmor kernel messages are blcoked from viewing from inside the container, and can only be seen on the host. Try running `dmesg` inside a lxd container, even as root.

File: features/environment.py [hunk 2]
```diff
@@ -498,22 +499,47 @@ def after_scenario(context, scenario):
...
-            if ("DENIED" in msg and "ubuntu_pro_" in msg)
-        ]
-    return apparmor_denied
+def _get_relevant_apparmor_logs(context):
+    if hasattr(context, "machines") and SUT in context.machines:
+        sut = context.machines[SUT]
+        if sut.cloud == "lxd-container":
+            # get apparmor DENIED messages from the host
+            with open("/var/log/syslog", "r") as syslog_fd:
+                syslog_messages = syslog_fd.readlines()
```
Comment (panlinux) [inline features/environment.py:508]: In the case of a lxd container, yes, we need to check for apparmor errors on the host's /var/log/syslog.

File: features/environment.py [hunk 3]
```diff
@@ -498,22 +499,47 @@ def after_scenario(context, scenario):
...
+                    and sut.instance.name in msg
+                )
+            ]
+            return apparmor_denied
+        else:
+            # get apparmor DENIED messages from the SUT
+            syslog_dest = os.path.join(
+                UA_TMP_DIR, "{}-syslog".format(sut.instance.name)
+            )
+            sut.instance.pull_file("/var/log/syslog", syslog_dest)
```
Comment (panlinux) [inline features/environment.py:524]: The syslog file inside a lxd container unfortunately won't contain the apparmor denials. Nor anything else inside the container that I could find. It's always on the host in this case :(

Comment (panlinux) [reply features/environment.py:524]: Ah, ok, in this case it's NOT a lxd container, so it's correct to grab it from the "inside".

File: features/environment.py [hunk 4]
```diff
@@ -498,22 +499,47 @@ def after_scenario(context, scenario):
...
+            # get apparmor DENIED messages from the host
+            with open("/var/log/syslog", "r") as syslog_fd:
+                syslog_messages = syslog_fd.readlines()
+            apparmor_denied = [
+                msg.strip()
+                for msg in syslog_messages
+                if (
+                    "DENIED" in msg
+                    and "ubuntu_pro_" in msg
+                    and sut.instance.name in msg
```
[the remark this answers is not in the document]
Comment (orndorffgrant) [reply features/environment.py:515]: Yep, it includes a timestamp

File: features/environment.py (hunk 1 above)
Comment (panlinux) [reply features/environment.py:532]: Ah, right, sorry for jumping to an incorrect conclusion

**Comment by lucasmoura** [comment]
CI errors seem unrelated to this change

**Review by panlinux — APPROVED** [review]

**Pull request merged**